services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api.py

```diff
@@ -576,5 +588,27 @@ async def update_scheduler_data_label(scheduler_data: SchedulerData) -> None:
                         raise _RetryError() from e
                     log.debug(
                         "Skip update for service '%s' which could not be found",
-                        scheduler_data.service_name,
+                        service_name,
                     )
+
+
+async def update_scheduler_data_label(scheduler_data: SchedulerData) -> None:
+    def _update_labels_in_spec(spec: dict) -> dict:
+        spec["Labels"][
+            DYNAMIC_SIDECAR_SCHEDULER_DATA_LABEL
+        ] = scheduler_data.as_label_data()
+        return spec
+
+    await _update_service_spec(
+        service_name=scheduler_data.service_name,
+        update_spec_data=_update_labels_in_spec,
```

Comment:
THOUGHT on the design:

IMO callables are a bad choice here. 
1. the caller can write a callable with side effects (e.g. exceptions or block clode)
2. the callables that you use here are not pure (i.e. the inputs are modified). The caller does not know that you are creating a deepcopy before calling.

The operation you want to do is very well define. You have a tree-like structure (i.e. dicts of dicts ...) and want to update some entries. For that you need two things: a) the path to the branch you want to update and b) the value to replace/append

e.g. pydantic solves it very nicely using something like [``copy(update=...)`` ](https://pydantic-docs.helpmanual.io/usage/exporting_models/#modelcopy).

Using an analogy I would get ``specs`` first and update the tree-like structure as
```python
updated_specs = copy_tree(
    specs,
    update={
        "Labels":{DYNAMIC_SIDECAR_SCHEDULER_DATA_LABEL: scheduler_data.as_label_data()},
        "TaskTemplate": {"Placement": "Constraints": [ f"node.id == {node_id}", ] }},
    deepcopy=True
)

```

to implement ``copy_tree`` we could build it using solid tools like https://toolz.readthedocs.io/en/latest/api.html#toolz.dicttoolz.update_in

Reply:
@Andrei Neagu: maybe have a look in [https://github.com/ITISFoundation/osparc-simcore/blob/775690fb7c1e5e20ddca67aab92d3fd8747ad8da/services/director-v2/src/simcore_service_director_v2/utils/dict_utils.py](https://github.com/ITISFoundation/osparc-simcore/blob/775690fb7c1e5e20ddca67aab92d3fd8747ad8da/services/director-v2/src/simcore_service_director_v2/utils/dict_utils.py)
This is what I use for merging dictionaries. [corresponding tests](https://github.com/ITISFoundation/osparc-simcore/blob/775690fb7c1e5e20ddca67aab92d3fd8747ad8da/services/director-v2/tests/unit/test_utils_dict_utils.py)

Reply:
Thanks for the tips. I also like the non callback approach better.